overlord/devicestate/devicemgr.go

```diff
@@ -2119,32 +2119,37 @@ func (m *DeviceManager) SystemAndGadgetAndEncryptionInfo(wantedSystemLabel strin
 	opts := &gadget.ValidationConstraints{
 		EncryptedData: encInfo.StorageSafety == asserts.StorageSafetyEncrypted,
 	}
-	if err := gadget.Validate(gadgetInfo, sys.Model, opts); err != nil {
+	if err := gadget.Validate(gadgetInfo, systemAndSnaps.Model, opts); err != nil {
 		return nil, nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
 	}
 
-	return sys, gadgetInfo, &encInfo, err
+	return systemAndSnaps.System, gadgetInfo, &encInfo, err
+}
+
+type systemAndEssentialSnaps struct {
+	*System
+	Seed            seed.Seed
+	InfosByType     map[snap.Type]*snap.Info
+	SeedSnapsByType map[snap.Type]*seed.Snap
 }
 
 // loadSystemAndEssentialSnaps loads information for the given label, which
 // includes system, gadget information, gadget and kernel snaps info,
 // and gadget and kernel seed snap info.
-func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (
-	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, error) {
-
+func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (*systemAndEssentialSnaps, error) {
```

Comment:
I wonder if this should take an option struct of some kind to control whether it returns the seed or not, I'm wondering this because seed is a relatively heavy object so memory wise dropping it earlier if not needed would be good?

Reply:
I'll address this in a followup PR.